File: src/goal_mode.py
"""
Goal Mode — режим достижения цели.

Раньше здесь был отдельный статичный планировщик: одним вызовом LLM
составлялся фиксированный список шагов (JSON), после чего шаги слепо
выполнялись один за другим — без реакции на промежуточные результаты,
без возможности сделать несколько инструментов ради одного шага, и с
мёртвым VERIFY_PROMPT, который никогда не вызывался.

Теперь Goal Mode — тонкая обвязка над тем же адаптивным циклом
инструментов, что и обычный chat(): модель сама решает, что делать
дальше, на основе РЕАЛЬНЫХ результатов уже выполненных шагов, может
использовать сколько угодно инструментов подряд, сама себя
перепланирует при неудаче, и продолжает пока не сочтёт цель
выполненной (или не кончится бюджет итераций). Список шагов для
отчёта и обучения skill_learner собирается постфактум из реально
выполненных вызовов инструментов (через _progress_hook), а не
придумывается заранее.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path
from typing import Callable


@dataclass
class GoalStep:
    step_num: int
    description: str
    tool: str = ""
    params: dict = field(default_factory=dict)
    status: str = "pending"  # pending, done, failed
    result: str = ""


@dataclass
class Goal:
    description: str
    steps: list[GoalStep] = field(default_factory=list)
    status: str = "planning"  # planning, executing, done, failed
    report: str = ""


class GoalModeAgent:
    """
    Режим цели — прогоняет цель через адаптивный цикл инструментов
    основного агента, наблюдая за каждым реальным вызовом инструмента,
    чтобы построить чек-лист/отчёт и накормить skill_learner.
    """

    # Системная рамка поверх обычного SYSTEM_PROMPT — не меняет формат
    # вызовов (<tool_call>...), а только говорит модели не
    # останавливаться после одного действия.
    GOAL_FRAMING = """ЗАДАЧА ВЫШЕ — это ЦЕЛЬ, а не разовый запрос.

Работай самостоятельно, шаг за шагом, вызывая столько инструментов
подряд, сколько нужно. Если для цели нужно сначала что-то изучить
(структуру проекта, существующий код, репозиторий) — сделай это
ПЕРЕД тем как писать/создавать что-либо, и используй увиденное.
Если какой-то шаг не удался — не сдавайся, попробуй скорректировать
подход и продолжай.

Останавливайся и пиши финальный текстовый ответ ТОЛЬКО когда цель
полностью выполнена (или ты уверен, что дальше двигаться нельзя).
Не проси подтверждения — действуй."""

    MAX_GOAL_ITERATIONS = 80  # цели обычно требуют больше шагов, чем обычный чат

    def __init__(self, agent, on_progress: Callable | None = None):
        self.agent = agent
        self.on_progress = on_progress or print
        self.current_goal: Goal | None = None

    def pursue(self, goal_text: str, save_report_to: str = "") -> str:
        """Выполнить цель от начала до конца через адаптивный цикл инструментов."""
        self.on_progress(f"\n🎯 Цель: {goal_text}")

        goal = Goal(description=goal_text, status="executing")
        self.current_goal = goal

        # Наблюдаем за каждым реальным вызовом инструмента внутри chat(),
        # чтобы построить чек-лист постфактум — без выдумывания шагов заранее.
        def _on_tool_executed(name: str, params: dict, result) -> None:
            step_num = len(goal.steps) + 1
            status = "done" if result.success else "failed"
            text_result = result.output if result.success else (result.error or "")
            goal.steps.append(GoalStep(
                step_num=step_num,
                description=f"{name}({', '.join(f'{k}={v!r}' for k, v in list(params.items())[:3])})",
                tool=name,
                params=params,
                status=status,
                result=text_result,
            ))
            icon = "✅" if result.success else "❌"
            self.on_progress(f"⚒️  Шаг {step_num}: {name} — {icon} {str(text_result)[:100]}")

        goal_prompt = f"{goal_text}\n\n{self.GOAL_FRAMING}"

        self.agent._progress_hook = _on_tool_executed
        try:
            final_text = self.agent.chat(
                goal_prompt,
                max_iterations=self.MAX_GOAL_ITERATIONS,
                force_all_tools=True,
            )
        except Exception as e:
            self.on_progress(f"⚠️  Ошибка при выполнении цели: {e}")
            final_text = f"Цель прервана ошибкой: {e}"
        finally:
            self.agent._progress_hook = None

        done = sum(1 for s in goal.steps if s.status == "done")
        failed = sum(1 for s in goal.steps if s.status == "failed")
        goal.status = "done" if failed == 0 else ("failed" if done == 0 else "done")

        report_lines = [
            f"## Отчёт по цели: {goal_text}\n",
            f"✅ Выполнено шагов: {done}",
            f"❌ Ошибок: {failed}\n",
        ]
        if goal.steps:
            report_lines.append("### Шаги:")
            for s in goal.steps:
                icon = "✅" if s.status == "done" else "❌"
                report_lines.append(f"{icon} {s.step_num}. {s.description}")
                if s.result:
                    report_lines.append(f"   Результат: {str(s.result)[:150]}")
            report_lines.append("")
        report_lines.append("### Итог модели:")
        report_lines.append(final_text or "(без финального текста)")

        report = "\n".join(report_lines)
        goal.report = report

        if save_report_to:
            Path(save_report_to).parent.mkdir(parents=True, exist_ok=True)
            Path(save_report_to).write_text(report, encoding="utf-8")
            self.on_progress(f"\n💾 Отчёт сохранён: {save_report_to}")

        self.on_progress(f"\n🏁 Цель завершена: {done} шагов выполнено, {failed} ошибок")
        return report

File: src/goal_mode.rs
//! Goal Mode — автономное выполнение целей.
//!
//! Режим, в котором агент получает высокоуровневую цель и самостоятельно
//! выполняет её, используя инструменты, пока не достигнет цели или
//! не исчерпает бюджет итераций.

use crate::Agent;
use serde_json::Value;
use std::path::PathBuf;

/// Шаг выполнения цели — реальный вызов инструмента.
#[derive(Debug, Clone)]
pub struct GoalStep {
    pub step_num: usize,
    pub description: String,
    pub tool: String,
    pub params: Value,
    pub status: GoalStepStatus,
    pub result: String,
}

/// Статус шага.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GoalStepStatus {
    Pending,
    Done,
    Failed,
}

/// Цель — описание и список выполненных шагов.
#[derive(Debug, Clone)]
pub struct Goal {
    pub description: String,
    pub steps: Vec<GoalStep>,
    pub status: GoalStatus,
    pub report: String,
}

/// Статус цели.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum GoalStatus {
    Planning,
    Executing,
    Done,
    Failed,
}

/// Агент в режиме цели.
pub struct GoalModeAgent<'a> {
    agent: &'a mut Agent,
    on_progress: Box<dyn Fn(&str) + Send + Sync>,
    current_goal: Option<Goal>,
}

impl<'a> GoalModeAgent<'a> {
    /// Создать новый GoalModeAgent.
    pub fn new(agent: &'a mut Agent) -> Self {
        Self {
            agent,
            // ИСПРАВЛЕНО: было println!() — тот же класс бага, что
            // eprintln! в Agent::chat() (main.rs, см. комментарий там):
            // в TUI-режиме прямая печать в stdout из фонового таска рвёт
            // рамки ratatui. Дефолт теперь пишет в файловый лог; REPL сам
            // подписывается через on_progress(), чтобы прогресс всё равно
            // доходил до пользователя — но безопасным путём (см. repl.rs).
            on_progress: Box::new(|msg| crate::logging_system::info(msg)),
            current_goal: None,
        }
    }

    /// Установить функцию обратного вызова для прогресса.
    pub fn on_progress<F>(&mut self, callback: F)
    where
        F: Fn(&str) + Send + Sync + 'static,
    {
        self.on_progress = Box::new(callback);
    }

    /// Выполнить цель.
    pub async fn pursue(&mut self, goal_text: &str, save_report_to: Option<&str>) -> String {
        (self.on_progress)(&format!("\n Цель: {}", goal_text));

        let mut goal = Goal {
            description: goal_text.to_string(),
            steps: Vec::new(),
            status: GoalStatus::Executing,
            report: String::new(),
        };

        self.current_goal = Some(goal.clone());

        let goal_framing = format!(
            "{}\n\n{}",
            goal_text,
            "ЗАДАЧА ВЫШЕ — это ЦЕЛЬ, а не разовый запрос.\n\n\
             Работай самостоятельно, шаг за шагом, вызывая столько инструментов\n\
             подряд, сколько нужно. Если для цели нужно сначала что-то изучить\n\
             (структуру проекта, существующий код, репозиторий) — сделай это\n\
             ПЕРЕД тем как писать/создавать что-либо, и используй увиденное.\n\
             Если какой-то шаг не удался — не сдавайся, попробуй скорректировать\n\
             подход и продолжай.\n\n\
             Останавливайся и пиши финальный текстовый ответ ТОЛЬКО когда цель\n\
             полностью выполнена (или ты уверен, что дальше двигаться нельзя).\n\
             Не проси подтверждения — действуй."
        );

        // Отслеживаем вызовы инструментов через канал: замыкание-хук должно
        // быть 'static (Agent хранит его как Box<dyn Fn + Send + Sync>, без
        // лайфтайма), поэтому вместо заимствования `&mut goal` изнутри хука
        // события просто отправляются в канал и разбираются здесь после
        // того как `chat()` полностью завершится.
        struct ToolEvent {
            name: String,
            params: Value,
            success: bool,
            output: String,
        }
        let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<ToolEvent>();

        let hook_tx = tx.clone();
        let hook = Box::new(move |name: &str, params: &Value, success: bool, output: &str| {
            let _ = hook_tx.send(ToolEvent {
                name: name.to_string(),
                params: params.clone(),
                success,
                output: output.to_string(),
            });
        });
        self.agent.set_tool_hook(Some(hook));

        let final_text = self.agent.chat(&goal_framing, 80, true).await;

        self.agent.set_tool_hook(None);
        drop(tx);

        while let Ok(event) = rx.try_recv() {
            let step_num = goal.steps.len() + 1;
            let status = if event.success {
                GoalStepStatus::Done
            } else {
                GoalStepStatus::Failed
            };
            let description = format!(
                "{}({})",
                event.name,
                event
                    .params
                    .as_object()
                    .map(|obj| {
                        obj.iter()
                            .take(3)
                            .map(|(k, v)| format!("{}={:?}", k, v))
                            .collect::<Vec<_>>()
                            .join(", ")
                    })
                    .unwrap_or_default()
            );
            let icon = if event.success { "✅" } else { "❌" };
            let truncated = if event.output.chars().count() > 100 {
                format!("{}…", crate::truncate_chars(&event.output, 100))
            } else {
                event.output.clone()
            };
            (self.on_progress)(&format!("⚒️  Шаг {}: {} — {} {}", step_num, event.name, icon, truncated));

            goal.steps.push(GoalStep {
                step_num,
                description,
                tool: event.name,
                params: event.params,
                status,
                result: event.output,
            });
        }

        let done = goal.steps.iter().filter(|s| s.status == GoalStepStatus::Done).count();
        let failed = goal.steps.iter().filter(|s| s.status == GoalStepStatus::Failed).count();
        goal.status = if failed == 0 {
            GoalStatus::Done
        } else if done == 0 {
            GoalStatus::Failed
        } else {
            GoalStatus::Done
        };

        let mut report_lines = Vec::new();
        report_lines.push(format!("## Отчёт по цели: {}\n", goal_text));
        report_lines.push(format!("✅ Выполнено шагов: {}", done));
        report_lines.push(format!("❌ Ошибок: {}\n", failed));

        if !goal.steps.is_empty() {
            report_lines.push("### Шаги:".to_string());
            for s in &goal.steps {
                let icon = if s.status == GoalStepStatus::Done { "✅" } else { "❌" };
                report_lines.push(format!("{} {}. {}", icon, s.step_num, s.description));
                if !s.result.is_empty() {
                    let truncated = if s.result.chars().count() > 150 {
                        format!("{}…", crate::truncate_chars(&s.result, 150))
                    } else {
                        s.result.clone()
                    };
                    report_lines.push(format!("   Результат: {}", truncated));
                }
            }
            report_lines.push(String::new());
        }

        report_lines.push("### Итог модели:".to_string());
        report_lines.push(final_text.clone());

        let report = report_lines.join("\n");
        goal.report = report.clone();
        self.current_goal = Some(goal);

        if let Some(path) = save_report_to {
            if let Some(parent) = PathBuf::from(path).parent() {
                let _ = std::fs::create_dir_all(parent);
            }
            let _ = std::fs::write(path, &report);
            (self.on_progress)(&format!("\n Отчёт сохранён: {}", path));
        }

        (self.on_progress)(&format!("\n Цель завершена: {} шагов выполнено, {} ошибок", done, failed));
        report
    }
}
